### PR TITLE
Fix required GraphQL resolver types

### PR DESCRIPTION
**Pull request by cpd67**
#### Description of what you did:
Related to #4548. We were having issues with the way strapi generated the graphql schema for our models; in particular, the way it handled non-nullable types. Because of ShadowCRUD, strapi generates queries for each model using the model's controller methods. The ones we were having trouble with were those for the `find` & `findOne` controller methods. Strapi did not mark the return type as required in the generated schema for those queries, and this caused issues when we used tools like graphql-codegen to generate types. 

If you use strapi's getstarted app as an example, and want to query for Restaurants, the queries for a single Restaurant & a list of Restaurants looks like this in the generated schema:

image

This causes issues with graphql-codegen when we write queries for all Restaurants, for example:
image

as this is what the type look like:
image

The list is nullable, but codegen is expecting otherwise. 

To solve this, we added changes to mark the return types as required in the type definitions file for the strapi-plugin-graphql package. This is what the queries for a single Restaurant & a list of Restaurants now look like in the schema:
image

And the types are generated correctly when using codegen:
image

### CODE REVIEW

File: packages/strapi-plugin-graphql/services/type-definitions.js [hunk 1]
```diff
@@ -378,7 +378,7 @@ const buildCollectionType = model => {
     if (actionExists(resolverOpts)) {
       _.merge(localSchema, {
         query: {
-          [`${singularName}(id: ID!)`]: model.globalId,
+          [`${singularName}(id: ID!)`]: `${model.globalId}!`,
```
Comment (martin-css) [inline packages/strapi-plugin-graphql/services/type-definitions.js:377]: I don't think it's valid to assume non-null here, a valid query may return null when asked for an ID that does not exist.

Agree with both the non-null constraints added to the plural query below.

**Comment by martin-css** [comment]
@cpd67, the lack of non-null constraints is also causing us issues with automatically generated Typescript definitions so would be good to see this get added.

Added a comment to the changes, for singular queries (which map to `findOne`) a null result is valid as the ID may refer to a non-existing object. However, agree with the changes to the plural queries.

**Comment by darron1217** [comment]
Does this should be included in major update?
I think it's a minor update(fix) for what it should have been. (Except for singular type definition)

**Comment by abdonrd** [comment]
> Hi @cpd67 The reason we haven't set those to required by default is to allow customization of the controllers without fearing a but if the user decides to return null. Doing this would actually be a breaking change in that sense. I will plan this for v4 :)

There is any ETA about this v4?

**Comment by FibHeap** [comment]
We encounter the same with codegen and typescript. The plural version should be 
``` [`${pluralName}(sort: String, limit: Int, start: Int, where: JSON)`]: `[${model.globalId}!]!`, ```

the single version should not be changed as @martin-css  described. 

Currently., it forces as to really write ugly code:
for a query like this:
```qgl
query clusters {
  clusters {
    id,
    header
    description
  }
}
```
we end up with the following code where every '?' seems to be meaningless...
image

Long story short we would also really appreciate a fix for this (maybe even before 4.0 😉)

**Comment by abdonrd** [comment]
@cpd67 could you update this PR with current `master`?

**Comment by alexandrebodin** [comment]
Hello there, v4 is out and has better management for required attributes. I'll go ahead and close this.  Thank you for contributing!

**Pull request closed without merging**